### DISCUSSION TITLE
Microsoft .NET to .NET Core Upgrade Assistant

### DISCUSSION THREAD

**Discussion by thabaum** [dnnsoftware/Dnn.Platform#5625] [General] [answered]
[Microsoft .NET Upgrade Assistant](https://dotnet.microsoft.com/en-us/platform/upgrade-assistant)

This is not the "Magic Wizard" we may see one day but anyone looking to upgrade this tool helps give steps on what to do to make the transition to .NET Core I found today I thought I would share.  Not sure if the Upgrade Assistant has been discussed as an option for helping in the event DNN Community decides to go to .NET Core?

```
Move your .NET Framework applications to .NET at your own pace with confidence.

Reduce time and difficulty modernizing codebases

The .NET Upgrade Assistant is a .NET global tool that helps you incrementally upgrade your .NET Framework-based Windows applications. Take advantage of performance gains, deployment flexibility, and innovation by upgrading to the modern .NET platform.

Significantly speed up your modernization efforts and get a head start on your journey to the cloud.
```

```
Multiple project types and languages supported

Whether it's an MVC web app or Windows desktop app, the .NET Upgrade Assistant can help. Currently, the tool supports the following project types:

    [ASP.NET MVC](https://learn.microsoft.com/dotnet/core/porting/upgrade-assistant-aspnetmvc?WT.mc_id=dotnet-35129-website)
    [Windows Forms](https://learn.microsoft.com/dotnet/core/porting/upgrade-assistant-winforms-framework?WT.mc_id=dotnet-35129-website)
    [Windows Presentation Foundation (WPF)](https://learn.microsoft.com/dotnet/core/porting/upgrade-assistant-wpf-framework?WT.mc_id=dotnet-35129-website)
    Console
    Libraries
    [UWP to Windows App SDK (WinUI)](https://learn.microsoft.com/dotnet/core/porting/upgrade-assistant-uwp-framework?WT.mc_id=dotnet-35129-website)
    [Xamarin.Forms to .NET MAUI](https://github.com/dotnet/upgrade-assistant/blob/main/docs/maui_support.md?WT.mc_id=dotnet-35129-website)

The tool supports C# and Visual Basic projects.
```

**Comment by sleupold** [comment]
unfortunately, DNN is a .Net WebForms application which is not supported by this tool.
https://learn.microsoft.com/en-us/dotnet/architecture/porting-existing-aspnet-apps/migrate-web-forms

**Comment by mitchelsellers** [comment, marked as answer]
As @sleupold the real issue is that System.Web, which is an underpinning requirement for DNN doesn't exist in .NET Core.

**Comment by thabaum** [comment]
Thanks for the feedback I will pass it along!

I will see what the plan is for the migration tool and if there is something cooking up for a future release that may help us here.  I asked if a plan for this type of migration fits the scope of the project and if so I will link the issue(s) related inside the github repo located here:

https://github.com/dotnet/upgrade-assistant/issues

**Reply by mitchelsellers** [reply]
Per Microsoft (https://learn.microsoft.com/en-us/dotnet/architecture/porting-existing-aspnet-apps/migrate-web-forms)

There is no plan to fix the lack of pathway from WebForms -> .NET Core

**Reply by thabaum** [reply]
squeaky wheels sometimes get oiled eventually...

"Web Forms will continue to be supported for quite some time. One option may be to keep this functionality in an ASP.NET 4.x app."

Would it be plausible to think we could do this?

Also I read further here:
https://fritzandfriends.github.io/BlazorWebFormsComponents/Migration/readme/

Forgive me its just hard to think all the time and energy and money invested moving from LAMP to C# pretty much the worst choice I ever made.  And to think they wont do it again to me makes me think about leaving C# even after all I have invested into learning .NET Core and Blazor stack.  I am a bit HOT over it all for years now and it's like a middle finger pointed up at all who can't afford it by MS.

For a guy that has been born in rags, I have devoted all my time into helping this project and building a company utilizing .NET technologies while supporting a lot of clients with MS products for a couple decades now small and medium sized organizations.

From that and Google burying my search results since I am not one of them... (having GoDaddy tell me I needed to do better SEO on my site pissed me off because I had #1 positions in DuckDuckGo, Bing, Yahoo... yet only DNN ToS page the one I did not even have listed on sitemap would show up 5 - 10 pages deep in Google search.  Its now #1 on google) to exposing that to get them to list my company as I had proof to back it up.  Now my site is listed but my tech is dated, all invested is difficult to maintain.  Google even sent my company checks to my state unclaimed funds twice for some reason I dont even know why, I dont do business with them... for LONG time.  Only thing I can say is I must be a threat to them... or was.  My plan was to work with all, but not sure I can with headwinds they produced such as changing things so our login systems in DNN would not work anymore and never go updated to support new OAuth which killed ALL my traffic.

Well now you have it, my tell all.  My motivation behind why I keep coming here.  I just want to run my business, not code.  I spent money on modules, support... and I just cant afford it anymore.   So I took a few online courses full stack developing Blazor apps of  a few different types from games to eCommerce and so on full stack to get a grasp on the entire stack.  I am not afraid of moving away from .NET, I just always supported local (MS is local to me and I produced community support for MS products locally for decades).

I have kept this to myself for very long time not spreading news like this.  I will overcome it all eventually and knowing what I am up against now has helped.

So either I am building a complete system to migrate from to Blazor (OR LAMP/JS/...) or migrating to Oqtane with help of building migration scripts for others to use as well dealing with the same headwinds I have.

With this all being said I am looking for a solution I can believe in working with over rest of my life and fast.  I have learned so many languages now over my time from pascal,C,C++,C#,java,python, php, mysql... they are all looking the same to me.

What is sad is my project I am so dedicated to is to help local communities and businesses of all sizes... and I kept getting shot down for one thing or another.  I have little resources and my family has taken the hit and remind me of it daily as I have invested so heavily here.

It's like taking your whole life and crumbling it up like a piece of paper and throwing it in the gutter almost.  And I just wont quit, I am determined to build this for our communities.  There is nothing like it as no one like me runs something like I am going to build eventually.  I am just trying to get the foundation laid out over 10 years now so I can get into the fun ideas I have to build on top if it.   And it is just sad, really sad thinking how I have put my family through such hardships investing in MS .NET Framework, however things happen for a reason I believe.  I believe there is a bigger mission to it all and it will all work out when needed most.  Somehow things hold together and I thank God for keeping me on track with moving forward no matter what.  From new rules to keep companies like mine from startup that I have had to hurdle from our latest governments on...  It's really messed up and that COVID bomb didn't help.  In fact Mitch, you and I was going to meetup if you remember... 

I know I have vented here a few times, but this is just feedback.  I dont hold grudges or look to get even, I am just looking to overcome the headwinds as once I do I will be able to share the success, unlock those who horde it.

**Comment by valadas** [comment]
@thabaum you seem to have a lot of frustrations with a lot of different things here in this very long post.

DNN is a mature platform that was built from the ground up on webforms, it also supports mvc and webapi. There are templates available for many different frontend technologies and it does not enforce anything on the developers and supports a very very wide variety of development patterns. Most of those that have been using the platform for a long time appreciate this.

Octane was started with _some_ of the same overall ideas as DNN but starting from scratch on today's new stacks. Maybe that would be a good solution for you if find that the benefits of .netCore outweigh the benefits of DNN ?
Orchard Core is another option.
All you learned about C# is not lost, C# has nothing at all to do with .netCore, you have not wasted any learning there, this will be useful in the 3 solutions above and in many other ways.

.NetFramework is not out of support and won`t be for quite some time. .Net Core on the other hand has very very short LTS cycles, it is a much faster-moving target. If we drop support for .NetFramework quicker than we need to, it would kill almost all 3rd party extensions and the platform with it.

DNN being open-source, nobody is getting paid to work on this and migrating to .Net Core is not a priority or a focus for those spending their free time working on and with the platform. Most need to support older extensions, keep maintaining older sites alive and secure, etc. That being said, some great efforts were put in place by such awesome individuals to get us part of our codebase on .netStandard and having dependency injection available in most places. So we are not refusing any efforts to get us closer to .netCore either but not at the cost of killing 3rd party extensions/themes, etc.

May I ask why you want .NetCore so much? There are only a handful of good reasons to have to migrate to .NetCore
- Microservices architecture, which you can already do, build your microservices in .netCore and just have DNN consume those. You already learned c# and you can use that talent for both.
- Better performance/scalability (but only if you really really have big workloads and lots of processing). If not then the major bottleneck would be database/file access which .netCore does not especially solve. Simple caching strategies will handle that on both .NetFramework or .NetCore
- Cross-platform support. This could potentially translate into a very small hosting cost savings but then your maintenance costs would go up exponentially with the very short LTS lifespan on .NetCore

>I am looking for a solution I can believe in working with over rest of my life

Well, 1st, maybe development is not your thing :) What is flagship today may very well be obsolete in a decade. 10 years ago I was the biggest javascript hater and loved webforms, telerik controls and using an insane amount of dependencies. A few years after I just wanted to kill webforms stuff everywhere and do React with javascript. A few years later now I do more typescript than anything and try to avoid dependencies like the plague. We may soon be just trying to explain stuff to robots and having to deal with AI integrations everywhere.
Secondly you would have wayyy more years of stability on .NetFramework than on .NetCore if building long-term solutions is your concern. .NET 6.0 (current LTS) will be out of support in November 2024, .NET 7 before that and .NET 8 is not even out. Sooo unless you are ready each year to invest time migrating solutions around, don't go the .NET Core way. By the way, those .NET Core end dates are set in stone. On the other hand, .NetFramework goes by the "Windows Components" support policy, this means it goes by the parent OS end of support date. Since it was a windows server 2022 component, it is at least supported until 2031. And that is not a date set in stone if windows 2023 or 2024 still ship with it, that date gets pushed even further. Microsoft would be crazy to not ship .NetFramework with windows anyway as it would kill soooo much of their enterprise licenses on windows.

As a final thought you may find this blog post of interest regarding DNN stance regarding a migration to .Net Core https://mitchelsellers.com/blog/article/the-technical-future-of-dnn it summarizes a lot of lengthy discussions the team had about this subject. Things may change in the distant future but right now we are focusing on improving DNN rather than migrating it.

**Reply by thabaum** [reply]
@valadas thanks for your kind thoughts.   I have spent a ton of time working on development and maybe your right, computers maybe not my thing in general I am beginning to learn.  Maybe I should be a carpenter or mechanic.  

I have been developing on .NET core and really I have learned to like it a lot compared to paste experiences I can develop rapidly things I have been looking to build for DNN now that I fully understand it's architecture.  In fact I was told by a MS employee to build from scratch and ditch any CMS as the best option.  That was about around .NET 3.  That thought has resonated in my head a lot lately as I am trying to help support.

So with all these thoughts racing around in my head decided to do both.  Build on top of Oqtane and build my entirely own solution on .NET Core.  

You are a great asset here and I really appreciate all the kind guidance you have given me over the years.  I suppose it was a bummer to see DNN not make DNN Core with migration scripts.    I really don't want to ditch this community as many of you have very good feedback over the years.  My hat's off to all of you keeping things alive here.

I wrote many times about issues with DNN since it was on Codeplex. I was thinking that I was in fact helping the community developers to learn that I really had to go in and fix the administrator test email SMTP issues myself and had to get very familiar with the code base for DNN which I reported since I started using it, but never got fixed until I fixed it.  I think it was one line of code maybe 2 that made the difference.  

I also fixed the issue with DNN Journal links issue which I still need to provide an acceptable PR for and I have not forgot about it, if someone wants to pickup where I left off I was leaving notes for anyone.  Someone could recompile my presented fix into a DNN journal module and install it if not interested in all the administration controls to turn it on and off for links created for journal entries which is needed to get accepted.  That feature worked for long time when I first started using DNN, then all sudden quit working.  I just found it hard with sites like Nasa, Yahoo and so on using this CMS that the community support which included entities like these and DNN Corp... could not give community fixes for these types of issues while they work with juiced up versions.  I had very high hopes for DNN in my mind, bubble is popped.

I found it sad that I had to fix this and no one cared to touch it (except community people like me), although it took a while for me to do it diving deep into the backend code of DNN green as can be in this area and fixed it.  DNN had a lot of internal issues I think kept it from growing into what it could of been.  But things happen for a reason, like you said @valadas  not all is lost here.  Keep up the inspiring words not many have that kind of touch.  At that time I was probably better at back-end development and I was working on my front-end UI/UX skills which I am still working on honing skills in this area.

I apologize but just found it hard to believe these issues were ignored.  I thought being a community person reporting all these issues helped out the core engineers fix them.  I found out if you want it done, you got to do it yourself.  Not everyone is a coder, not everyone has time to dive that deep into a code base.  I had to show the fix to prove the issue more or less.  When you see big global meetings on an Open Source project I suppose I was led to believe that this had some backing that would support fixing the bugs like these as needed.  It was really a pretty fairly easy fix...  When I get excuses like "we are all community developers that don't get paid" it reminds me of when you call a business and the employee says "i just work here I am not this company" when actually them being an employee makes them part of that company and they should really care to address the issues presented by a client/customer if you know what I am saying.  It's not 100% the same here I got it, but the flip/flop of DNN's (DotNetNukes) history with all kinds of big money floating around it dictating its outcome while everyone is "not paid".  When DotNetNuke bought out Active Forums and all the social modules I was expecting a big boost in development in this area, instead the project in a way died in this area.  Only option was $10K/year+...

At that time I was probably better at back-end development and I was working on my front-end UI/UX skills which I am still working on honing skills in this area.

So maybe a full blown creation now that I have had to invest so much time, money and effort in this area is my best solution not letting my skills I have go to waste so to speak in regards to C#.  Also decoupling myself from the open source worries of breaking changes in regards to other module developers.

Just bottling up all that experience I believe is extremely valuable to someone that maybe in the same boat.

@mitchelsellers @valadas @sleupold @bdukes @Timo-Breumelhof @david-poindexter + all others, thanks again for all the years helping me through the world of online development in the Microsoft ecosystem and the DotNetNuke platform.  I really never thought it would end like this, being a die-hard i really tried to hang tough through the thick of it... but it's time to cut this line.

And once again for all the years of putting up with my critical feedback I can't thank you all enough, you all really ROCK!  

Good luck with everything in the future with DNN and life in general!  As a final contribution I may share my migration scripts to Oqtane (and create all the base modules DNN provides for Oqtane) in a repo if I decide that is my direction for future projects I have running on DNN.  It wont be a fix all solution, but I will see what I can do for the common modules like the HTML module in DNN to migrate to Oqtane HTML module and maybe a few others that basic site owners can use to help as a migration path to if desired.

I am going to just ditch everything I had including all my vendor bought modules and create my own.  Write it all off as lessons well learned.  

@iJungleboy has created https://blazor-cms.org/ as well but I am not sure if there are migration scripts for DNN to Blazor CMS or if there is plans for them?  Thought I would toss a plug out for your CMS since it is also an option anyone looking for something other than Orchard Core.  Since you have heavy ties here and in Oqtane

**Comment by valadas** [comment]
>I found it sad that I had to fix this and no one cared to touch it

You need to understand again that because you created an issue does not mean it will be handled quick or at all. The list of issues here are to inform potential contributors. I do tackle issues I care about in a particular order, security, then things that affect me or my clients, then anything I know a sponsor of mine needs, then from oldest to newest. Each contributor is free to prioritize any other way too. So, it's not that nobody cares to touch your particular issue, just nobody got to it by the time you did. And it was obviously important for you so you submitted a PR. That is totally fine, don't blame the community for not handling your issue fast enough. During that time hundreds of other issues probably got handled.

>When you see big global meetings on an Open Source project I suppose I was led to believe that this had some backing that would support fixing the bugs like these as needed

We have a few "teams", for instance:
- .Net Foundation: Their responsibility here is to ensure that DNN stays open-source, and they also provide some free resources like Azure for our builds, a system to manage contribution license agreements, they can help promote, etc. They do not send the community or any developer any money or other compensation.
- The Technology Advisory Group is a group of people that have been primarily focused on DNN for many many years. They have a wide knowledge of both the current and historical Dnn overall architecture and take global decisions about the future of this open-source project. This team is not paid, it is not a company, etc. It is just simply a free coordinated effort to give DNN some guidance. That blog post I link earlier about the future of DNN was discussed without that team to make sure that group agreed with this plan before posting. That team also reacts to security reports and architect plans to fix them in the best way.
- The Approvers group are people vetted to know the platform well enough and have enough experience to review code contributions. They also meet about every 2 weeks, they are also not paid, they just want to help out planning new releases and prevent the platform from getting broken by bad contributions. They also do something very important which is to respond to each Pull Request in a timely manner so that people who spent countless hours submitting code don't feel ignored and get some help getting the code right and merged in. Don't quote me on the exact numbers but before that team creation I believe this repository had something like 200 pending very old PRs and nearly 1000 issues. 
- There are a bunch of other meetings going around like Saturday open-source co-coding sessions, SouthernFried DNN, 2sxc meetups, TADUG was a popular one, etc. Those are just individuals steaming because they love talking about the platform, they are not getting paid by anyone to do this. At best, they get 100 views, which is not even enough to make any money out of youtube or other platforms. They simply do that because they love doing it.
- Now things like the DNN Association and DNN Connect are actual entities, they are non-profit though, they were created to manage DNN events like DNNConnect and DNNSummit.

Now if we are talking EVOQ, this is a commercial project and developers that work on it are getting paid. They can pay them because they sell EVOQ. it is based on DNN Platform and they often contribute back bug fixes to this repository when it is appropriate to fix it for both DNN and EVOQ. Now if they build something new or fix something that does not exist in the platform, they obviously won't make that available for free. They have their own separate support channels and they prioritize what they do based on what their clients need. So for instance, if you were an EVOQ customer, you could contact their support channels and then have some reason to be mad about not taking care of your issue because you are paying for such things.

> I am going to just ditch everything I had including all my vendor bought modules and create my own.

Well that's the exact reason we can't just break the ecosystem to migrate to .NetCore. Making that move would break all 3rd-party extensions. So if you don't have to deal with existing sites and value .NetCore more than the maturity of DNN. By all means, change platforms to what you like.

And as a final thought, still don't know why you want .NET Core

**Reply by thabaum** [reply]
`You need to understand again that because you created an issue does not mean it will be handled quick or at all. The list of issues here are to inform potential contributors.`

https://github.com/dnnsoftware/Dnn.Platform/pull/3237

One of my first contributions after reporting this for years...  I did not even use visual studio or any help other than old fashion look at code and study it.  In fact I did not even know how to test it.  I was flying blind if you remember (or maybe I was quiet about it) but it got accepted and I was blown away excited when it did.  It was just very hard to setup a new portal site when it sent emails from the wrong email account when testing email configurations setting up multiple sites in one DNN installation which is what attracted me to DNN.  Again I was hoping to focus on CSS, maybe some adjustments here or there.  It was over 10 years ago... reporting it over and over each time trying to find help with the situation.  That issue plagued me and others I believe a very long time as you can find so many posts about it dating way back.

https://dotnetnuke.codeplex.com/ no longer exists but I did report it I believe when it existed here too.

Look up SMTP issues and DNN you will see more than just me, it was a major issue being ignored it felt.

https://github.com/dnnsoftware/Dnn.Platform/issues/3142

I guess I was not your normal code jockey the development community was used to.

As for Developer modules, if they are holding back a project like this, it is hard to want to be included in it.  I was hoping the community would be working together to move things forward.  I was caught in a blame game pickle between module developer and DNN what was going on with CSS styles.

Follow https://github.com/dnnsoftware/Dnn.Platform/issues/3578
https://github.com/dnnsoftware/Dnn.Platform/issues/3611

Why .NET Core?  

https://www.netsolutions.com/insights/net-core-vs-net-framework/

Who knows maybe Typescript or Python is the way to go... but that article sums up some reasons.

Not trying to hate here, I know big money ties hands.  It is out of our control.  We get whatever gets trickled down, then big change comes when you catch up.

|https://github.com/dnnsoftware/Dnn.Platform/issues/2257
https://github.com/dnnsoftware/Dnn.Platform/issues/2981
https://github.com/dnnsoftware/Dnn.Platform/issues/3035

More love the community could use above, I would love to complete all of this but I just cant see doing it.  Again cutting the line here as it has pulled my boat backwards...

**Comment by mitchelsellers** [comment]
I was out of the office when all of this came in, but I'll just add a little bit more here as well.

Technology selection is a very delicate undertaking and is riddled with risks, pros/cons, and adjustments.

Microsoft is for sure moving in the direction of .NET Core, much larger percentage of my own business is .NET Core based than otherwise right now, but that doesn't mean that .NET Framework isn't neeed, or even the RIGHT solution.

.NET Framework has a much different support pattern, it has a lot more adoption base and history.  It is possible to set-and-forget with minimal risks.  It also can perform well and handle various workloads.

.NET Core is the new shiny, it is faster, but it requires a complete shift to the nature of what you are doing, and how you support business.  

I personally find that when it comes to DNN usage, I favor a hybrid approach, which I've blogged about and talked about in the past.

The fact of the matter is that there is NOT a single way to transition from WebForms -> Anything else.  This means, regardless of if any future "DotNetNuke" like product exists, you will be 100% required to re-do major portions of your codebase.  For most people the juice isn't worth the squeeze to do it.  Especially considering .NET 4.8 is supported until 2032 at least as-is.

Any platform that you use, .NET Core, .NET Framework, Angular, React, Vue, Bootstrap, etc comes with a set of risks. Your pathway forward will be often dictated by the direction of that product.  Early adopters for example of AngularJS (1.x) got a real taste of the bad with the major changes that had to happen when you move forward.  For those of us that are responsible for technology stack selection these are things that we have to address every day.

As to your notes about issues or ignored issues.  Our community is strong, there was a history (Pre 2019ish) where the community was not fully driving direction, managing pull requests, etc.  Since then however we try to be inclusive, but we do ONLY accept changes that are not going to have a negative impact on the community or the product.  As such, if a solution will result in a one-sided fix, or might break something from elsewhere it is possible we will need to request changes or adjustments.  Often times, those come in from the contributors, but again, we may not be able to drill 40+ hours into trying to help get a partial fix over the hump.

**Reply by thabaum** [reply]
Great point of "set-and-forget" since little to no changes are coming over next 10 years to enhance it, but security and maybe some performance updates.  Set-and-forget reminds me of the new AI weapons being produced specially designed for pubic influence and warfare.

I had thought heavily of the Hybrid approach since .NET core was introduced as you can do this with many different technology stacks for different purposes.  I just can't handle all the other issues that come with DNN's Framework for some projects.  I was hoping some new light may shine this way to migrate easier to .NET Core.

It takes teamwork to make the dream work they say, and DNN team kept this project alive all these years through all these types of tough discussions which is admirable.  Best hope and wishes to all.

Thanks again @mitchelsellers !
